arrows/gdal/image_io.cxx

```diff
@@ -0,0 +1,241 @@
+  MAP_METADATA_COEFF( SAMP_NUM_COEFF, COL_NUM_COEFF )
+  MAP_METADATA_COEFF( SAMP_DEN_COEFF, COL_DEN_COEFF )
+
+#undef MAP_METADATA_SCALAR
+#undef MAP_METADATA_COEFF
+}
+
+/// Constructor
+image_io
+::image_io()
+{
+}
+
+/// Destructor
+image_io
+::~image_io()
+{
+}
+
+/// Load image image from the file
+/**
+ * \param filename the path to the file the load
+ * \returns an image container refering to the loaded image
+ */
+vital::image_container_sptr
+image_io
+::load_(const std::string& filename) const
+{
+  GDALAllRegister();
+
+  std::shared_ptr<GDALDataset> gdalDataset;
+
+  gdalDataset.reset( static_cast<GDALDataset*>(
+    GDALOpen( filename.c_str(), GA_ReadOnly ) ) );
+
+  vital::image img;
+
+  if ( !gdalDataset )
+  {
+    VITAL_THROW( vital::invalid_file, filename, "GDAL could not load file.");
+  }
+  else
+  {
+    // Get the size of the image and the number of bands
+    auto imgWidth = gdalDataset->GetRasterXSize();
+    auto imgHeight = gdalDataset->GetRasterYSize();
+    auto imgDepth = gdalDataset->GetRasterCount();
+
+    // Create a new vital image based on the GDAL raster type. For now just
+    // load bands directly into channels in vital::image.
+    // TODO: deal or provide warning if bands have different types.
+    auto bandType = gdalDataset->GetRasterBand(1)->GetRasterDataType();
+    switch (bandType)
+    {
+      case (GDT_Byte):
+      {
+        img = vital::image_of<vital::byte>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_UInt16):
+      {
+        img = vital::image_of<uint16_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Int16):
+      {
+        img = vital::image_of<int16_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_UInt32):
+      {
+        img = vital::image_of<uint32_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Int32):
+      {
+        img = vital::image_of<int32_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Float32):
+      {
+        img = vital::image_of<float>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Float64):
+      {
+        img = vital::image_of<double>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      default:
+      {
+        std::stringstream ss;
+        ss << "kwiver::arrows::gdal::image_io::load(): "
+           << "Unknown or unsupported pixal type: "
+           << GDALGetDataTypeName(bandType);
+        VITAL_THROW( vital::image_type_mismatch_exception, ss.str() );
+        break;
+      }
+    }
+
+    // Loop over bands and copy data
+    CPLErr err;
+    for (int i = 1; i <= imgDepth; ++i)
+    {
+      GDALRasterBand* band = gdalDataset->GetRasterBand(i);
+      bandType = band->GetRasterDataType();
+      err = band->RasterIO(GF_Read, 0, 0, imgWidth, imgHeight,
+        static_cast<void*>(
+          reinterpret_cast<GByte*>(img.first_pixel()) + (i-1)*img.d_step()),
```

Comment:
Since your actual image data type may not be bytes, I'm wondering if you need to multiply` img.d_step()` by `img.pixel_traits().num_bytes`